detect_test.go

```diff
@@ -0,0 +1,233 @@
+				Expect(err).To(MatchError(errors.New("some error")))
+			})
+		})
+
+		context("when $COMPOSER_VENDOR_DIR is not underneath the project root", func() {
+			invalidPaths := []string{
+				"/usr/vendor",
+				"../usr/vendor",
+				"./../vendor",
+			}
+
+			it("does not require or provide anything for invalidPath", func() {
+				for _, invalidPath := range invalidPaths {
+					Expect(os.Setenv("COMPOSER_VENDOR_DIR", invalidPath))
+					_, err := detect(packit.DetectContext{WorkingDir: workingDir})
+					Expect(err).To(MatchError(packit.Fail.WithMessage("COMPOSER_VENDOR_DIR must be a relative path underneath the project root")))
+				}
+			})
+		})
+	})
+
+	context("when $COMPOSER is set", func() {
+		it.Before(func() {
+			Expect(os.Setenv("COMPOSER", "other/location/composer.json")).ToNot(HaveOccurred())
+		})
+
+		context("when $COMPOSER points to an existing file", func() {
+			it.Before(func() {
+				Expect(os.Mkdir(filepath.Join(workingDir, "other"), os.ModeDir|os.ModePerm)).NotTo(HaveOccurred())
+				Expect(os.Mkdir(filepath.Join(workingDir, "other", "location"), os.ModeDir|os.ModePerm)).NotTo(HaveOccurred())
+				Expect(ioutil.WriteFile(filepath.Join(workingDir, "other", "location", "composer.json"), []byte("{}"), os.ModePerm)).NotTo(HaveOccurred())
+			})
+
+			it(`requires "composer" and "php" and provides "composer-packages"`, func() {
+				detectResult, err := detect(packit.DetectContext{WorkingDir: workingDir})
+				Expect(err).NotTo(HaveOccurred())
+
+				Expect(detectResult.Plan).To(Equal(packit.BuildPlan{
+					Provides: []packit.BuildPlanProvision{
+						{
+							Name: composer.ComposerPackagesDependency,
+						},
+					},
+					Requires: []packit.BuildPlanRequirement{
+						{
+							Name: "composer",
+							Metadata: composer.BuildPlanMetadata{
+								Build: true,
+							},
+						},
+						{
+							Name: "php",
+							Metadata: composer.BuildPlanMetadata{
+								Build: true,
+							},
+						},
+					},
+				}))
+
+				Expect(phpVersionResolver.ResolveCall.Receives.ComposerJsonPath).To(Equal(filepath.Join(workingDir, "other", "location", "composer.json")))
+				Expect(phpVersionResolver.ResolveCall.Receives.ComposerLockPath).To(Equal(filepath.Join(workingDir, "other", "location", "composer.lock")))
+			})
+
+			context("when composer.lock is not present as a sibling of composer.json", func() {
+				it.Before(func() {
+					Expect(ioutil.WriteFile(filepath.Join(workingDir, "composer.lock"), []byte("{}"), 0644)).NotTo(HaveOccurred())
+				})
+
+				it("will log a warning", func() {
+					_, err := detect(packit.DetectContext{WorkingDir: workingDir})
+					Expect(err).NotTo(HaveOccurred())
+
+					Expect(buffer).To(ContainLines("WARNING: Include a 'composer.lock' file with your application! This will make sure the exact same version of dependencies are used when you build. It will also enable caching of your dependency layer."))
+				})
+			})
+		})
```

Comment:
I'm confused why this is the behaviour (a warning about including a `composer.lock`) when the `composer.lock` exists

Reply:
The `composer.lock` file does exist, but not as a "sibling" of `composer.json`. The surrounding `context` specifies a different location for `composer.json`: